document/encoding/msgpack/encoding_test.go

```diff
@@ -0,0 +1,300 @@
+			`{"age": 10, "name": "john"}`,
+		},
+		{
+			"Map",
+			userMapDoc,
+			`{"age": 10, "name": "john"}`,
+		},
+		{
+			"Nested Document",
+			document.NewFieldBuffer().
+				Add("age", document.NewInt64Value(10)).
+				Add("name", document.NewTextValue("john")).
+				Add("address", document.NewDocumentValue(addressMapDoc)),
+			`{"age": 10, "name": "john", "address": {"city": "Ajaccio", "country": "France"}}`,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			enc, err := EncodeDocument(test.d)
+			require.NoError(t, err)
+			var buf bytes.Buffer
+			err = document.ToJSON(&buf, DecodeDocument(enc))
+			require.NoError(t, err)
+			require.JSONEq(t, test.expected, buf.String())
+		})
+	}
+}
+
+func TestDecodeDocument(t *testing.T) {
+	mapDoc, err := document.NewFromMap(map[string]string{
+		"city":    "Ajaccio",
+		"country": "France",
+	})
+	require.NoError(t, err)
+
+	doc := document.NewFieldBuffer().
+		Add("age", document.NewInt64Value(10)).
+		Add("name", document.NewTextValue("john")).
+		Add("address", document.NewDocumentValue(mapDoc))
+
+	data, err := EncodeDocument(doc)
+	require.NoError(t, err)
+
+	ec := DecodeDocument(data)
+	v, err := ec.GetByField("age")
+	require.NoError(t, err)
+	require.Equal(t, document.NewInt64Value(10), v)
+	v, err = ec.GetByField("address")
+	require.NoError(t, err)
+	var expected, actual bytes.Buffer
+	err = document.ToJSON(&expected, document.NewFieldBuffer().Add("address", document.NewDocumentValue(mapDoc)))
+	require.NoError(t, err)
+	err = document.ToJSON(&actual, document.NewFieldBuffer().Add("address", v))
+	require.NoError(t, err)
+	require.JSONEq(t, expected.String(), actual.String())
+
+	var i int
+	err = ec.Iterate(func(f string, v document.Value) error {
+		switch f {
+		case "age":
+			require.Equal(t, document.NewInt64Value(10), v)
+		case "address":
+			var expected, actual bytes.Buffer
+			err = document.ToJSON(&expected, document.NewFieldBuffer().Add("address", document.NewDocumentValue(mapDoc)))
+			require.NoError(t, err)
+			err = document.ToJSON(&actual, document.NewFieldBuffer().Add(f, v))
+			require.NoError(t, err)
+			require.JSONEq(t, expected.String(), actual.String())
+		case "name":
+			require.Equal(t, document.NewTextValue("john"), v)
+		}
+		i++
+		return nil
+	})
+	require.NoError(t, err)
+	require.Equal(t, 3, i)
+}
+
+func TestEncodeArray(t *testing.T) {
+	mapDoc, err := document.NewFromMap(map[string]string{
+		"city":    "Ajaccio",
+		"country": "France",
+	})
+	require.NoError(t, err)
+
+	tests := []struct {
+		name     string
+		a        document.Array
+		expected string
+	}{
+		{
+			"Complex array",
+			document.NewValueBuffer().
+				Append(document.NewInt64Value(10)).
+				Append(document.NewTextValue("john")).
+				Append(document.NewDocumentValue(mapDoc)).
+				Append(document.NewArrayValue(document.NewValueBuffer().Append(document.NewInt64Value(11)))),
+			`[10, "john", {"city": "Ajaccio", "country": "France"}, [11]]` + "\n",
```

Comment:
I don't remember why there is a `\n` at the end?

Reply:
It's because backquotes don't support `\n`, it's just a convenient way of adding a new line without actually writing one